### PR TITLE
[c++] fix Booster.refit() segfault with linear_tree on single-class data

### LINKED ISSUE #6792: [python-package] `refit()` segfaults when label only contains 1 class
## Description

Using the Python package, when fitting a binary classification model with only categorical features and using linear trees, `refit()` segfaults if `label` has only 1 class.

## Reproducible example

```python
import lightgbm as lgb
import numpy as np

X = np.random.randint(low=0, high=10, size=(1_000, 5))
y = np.ones(shape=(X.shape[0],))

dtrain = lgb.Dataset(
  data=X,
  label=y,
  categorical_feature=list(range(X.shape[1]))
)

bst = lgb.train(
  train_set=dtrain,
  params={
    "objective": "binary",
    "num_leaves": 7,
    "linear_tree": True
  },
  num_boost_round=2
).refit(X, label=y, categorical_feature=list(range(X.shape[1])))
```

Logs from `refit()`:

```text
[LightGBM] [Warning] Contains only one class
[LightGBM] [Info] Number of positive: 1000, number of negative: 0
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000032 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 55
[LightGBM] [Info] Number of data points in the train set: 1000, number of used features: 5
Segmentation fault: 11
```

Training the same model but without `linear_tree=True`, `refit()` does not segfault or otherwise raise an exception... it just trains a stub model, as expected.

```python
bst = lgb.train(
  train_set=dtrain,
  params={
    "objective": "binary",
    "num_leaves": 7,
  },
  num_boost_round=2
)
bst.refit(X, label=y, categorical_feature=list(range(X.shape[1])))
```

Logs from `refit()`:

```text
[LightGBM] [Warning] Contains only one class
[LightGBM] [Info] Number of positive: 1000, number of negative: 0
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000035 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 55
[LightGBM] [Info] Number of data points in the train set: 1000, number of used features: 5
<lightgbm.basic.Booster object at 0x152a19690>
```

Tree content:

```python
bst.dump_model()["tree_info"]
```

```json
[{"tree_index": 0, "num_leaves": 1, "num_cat": 0, "shrinkage": 1, "tree_structure": {"leaf_value": 34.53957599234088, "leaf_count": 1000}}]
```

## Environment info

LightGBM version or commit hash: https://github.com/microsoft/LightGBM/commit/226e7f7de8358161072524f4f91ed8366a1dead6

Command(s) you used to install LightGBM

```shell
cmake -B build -S . -DUSE_OPENMP=OFF
cmake --build build --target _lightgbm -j4
sh build-python.sh install --precompile
```

## Additional Comments

Fitting a model with a single class is maybe not that useful of a case for LightGBM, but still... LightGBM should **never** do something that results in process-crashing segmentation fault.

And these sorts of situations can occur unintentionally, for example in AutoML systems where random subsets of training data are selected.

### PR DESCRIPTION
**Pull request by FranciscoRMendes**
Fixes the segfault in Booster.refit() reported in #6792.
  
  When refitting a linear_tree=True model on data whose label has a single class, some features end up constant in that subset. LightGBM filters out constant features during Dataset construction, so Dataset.InnerFeatureIndex() returns -1 for them.
  CalculateLinear was then calling FeatureBinMapper(-1) — an invalid index — which caused the segfault.

  The fix skips features where InnerFeatureIndex() returns -1 in CalculateLinear. I added a leaf_old_coeff_idx vector to track each surviving feature's position in the stored coefficient array, so the decay blend (decay_rate * old_coeff + ...) still reads the
  right slot when features are skipped.

  I have tested this by running test_refit_linear_tree_single_class_no_segfault which confirms refitting on single-class data no longer crashes.

  Closes #6792

### CODE REVIEW

**Review by jameslamb — CHANGES_REQUESTED** [review]
> _Bug 2: Spurious UserWarning about categorical_feature keyword in params_

We already have a fix for #6793 in #7124, please remove changes related to that from this PR. And if you're curious, please do look at the changeset in that PR... quite a lot more than just removing the parameters is needed.

**Comment by FranciscoRMendes** [comment]
> > _Bug 2: Spurious UserWarning about categorical_feature keyword in params_
> 
> We already have a fix for #6793 in #7124, please remove changes related to that from this PR. And if you're curious, please do look at the changeset in that PR... quite a lot more than just removing the parameters is needed.

oops my bad. I thought this would be an easy fix for me, turns out I was wrong.

**Comment by FranciscoRMendes** [comment]
removed changes for the other unrelated bug

File: python-package/lightgbm/basic.py [hunk 1]
```diff
@@ -3115,7 +3115,7 @@ def set_weight(
         # Check if the weight contains values other than one
         if weight is not None:
             if _is_pyarrow_array(weight):
-                if pa_compute.all(pa_compute.equal(weight, 1)).as_py():
+                if pa_compute.all(pa_compute.equal(weight, 1)).as_py():  # type: ignore[attr-defined]
```
Comment (jameslamb) [inline python-package/lightgbm/basic.py]:
```suggestion
```

Please remove this and update this branch, this was fixed in #7236

And in the future, when you find yourself making changes that seem unrelated to the purpose of the PR, please consider proactively writing a small comment explaining them. Including the relevant warnings / error messages in plaintext is also helpful, because it allows others to find similar issues from GitHub search.

For this case, I did that here: #7238

Doing that before maintainers have to ask "wait, what is this about?" cuts out a review cycle, which helps us all work a bit more efficiently.

**Review by jameslamb — CHANGES_REQUESTED** [review]
Just one quick comment, will review more thoroughly when I can.

**Comment by FranciscoRMendes** [comment]
I did three things, 
  1. Guard against -1 feature index (if (feat < 0) continue;) — this is the direct crash fix. When a feature is constant in the refit data, InnerFeatureIndex() returns -1, and passing that to FeatureBinMapper() was the segfault.
  2. leaf_old_coeff_idx vector — tracks each surviving feature's original position in the raw feature list. This is necessary because once you start skipping features, position i in the new feature list no longer matches position i in the old coefficient array.
  3. Decay-blend index fix — uses old_coeff_idx[i] instead of i when blending old and new coefficients, so the right old coefficient is read even when some features were skipped.

  Without (2) and (3), the guard in (1) would fix the crash but silently use the wrong old coefficients during refit.

**Comment by BelixRogner** [comment]
👋 Thanks @FranciscoRMendes — adopted into [**ExaBoost**](https://github.com/BelixRogner/ExaBoost), a contribution-friendly CUDA-focused LightGBM fork. Merged as [`6f0066ff`](https://github.com/BelixRogner/ExaBoost/commit/6f0066ff).

Clean null/OOB crash fix in `CalculateLinear()` — exactly the class of host-side deref bug we care about. 🚀
